Add support to set clusterflags through API and cli

obnoxxx opened the pull request:
It is useful/needed to be able to change the cluster File and Block flags after cluster creation.

phlogistonjohn commented on client/cli/go/cmds/cluster.go:
```diff
@@ -43,10 +47,22 @@ func init() {
 			"\n\tvolumes on this cluster.")
 
+	clusterSetFlagsCommand.Flags().StringVar(&cl_block_str, "block", "",
+		"\n\tOptional: Control the possibility of creating block volumes"+
```
Perhaps "Control the possibility of" is more succinctly written as "Allow" or "Permit"

obnoxxx replied:
Actually, thinking about it, the text is not *that* wrong:

The `--block` option lets you specify whether or not it should be possible to create block volumes on the cluster. It lets us allow or forbid (i.e. control) that possibility. That's where I was coming from.

I still agree that it's not worded too well. Thinking how to frame it better...

phlogistonjohn commented on client/cli/go/cmds/cluster.go:
```diff
@@ -96,6 +112,60 @@ var clusterCreateCommand = &cobra.Command{
+		}
+
+		req := &api.ClusterSetFlagsRequest{}
+
```
Do we care if they run the command with neither --block=<x> nor --file-<x> ?
Right now, it would be a no-op, but a no-op that "writes" to the server. Maybe printing an error (or a warning) would be better to provide feedback to the user?

obnoxxx replied:
done. heketi-cli cluster setflags now expects at least one of --file or --block.

obnoxxx commented:
Branch updated.
* based on top of new PR #971 for manpage cleanup
* added manpage section for setflags
* added setflags operation to python client and added python-test
* heketi-cli cluster setflags now requires at least one of --file or --block to be specified

TODO:
* Find out whether introduction of ClusterFlags is an api violation that we can't do (look at external consumers kubernetes in-tree and external provisioners), and ether remove the patch or complete it to cover the functional tests....

obnoxxx commented:
rebased over merged PR #966 which changed the manpage

obnoxxx commented:
Fixed up the use of Block/File flags as literal struct member in the functional test cases.

@phlogistonjohn for the question about whether to do this change at all: This does not seem to be used outside. After all, we don't do cluster operations in our provisioners -- they only do volume operations.

obnoxxx commented:
Updates:
* rebased on top of #971 which is rebased on latest master.
* Added tests for the api server and api client.
* Updated the API documentation for the new call.
* updated the API doc for the block/file flags in the create cluster call.

Now this should be complete.
According to discussion with @phlogistonjohn, we should be good to do the api normalization with the ClusterFlags struct.

phlogistonjohn commented on client/api/python/test/unit/test_client.py:
```diff
@@ -42,6 +42,19 @@ def test_cluster(self):
+        info = c.cluster_info(cluster['id'])
+        self.assertEqual(info['id'], cluster['id'])
+        self.assertEqual(False, info['block'])
+        self.assertEqual(True, info['file'])
```
Small nit. Python unittest has assertTrue and assertFalse methods that are more appropriate for testing booleans. However, since the existing test code is already doing this all over the place (and some even more uglier stuff) I don't think its worth complaining too much about.

obnoxxx replied:
fixed, thanks

phlogistonjohn commented on doc/api/api.md:
````diff
@@ -224,6 +224,26 @@ Heketi is able to manage multiple GlusterFS clusters, each composed of a set of
 }
 ```
 
+### Set Cluster Flags
````
This looks great. One small thing is that there's no corresponding link at the top of the document to this section.

obnoxxx replied:
fixed, thanks

obnoxxx commented:
rebased on current master after #971 has been merged.

obnoxxx commented on client/api/go-client/cluster.go:
```diff
@@ -62,6 +62,42 @@ func (c *Client) ClusterCreate(request *api.ClusterCreateRequest) (*api.ClusterI
+	if r.StatusCode != http.StatusOK {
+		return utils.GetErrorFromResponse(r)
+	}
+
```
Does that mean that `ClusterDelete()` also needs `r.Body.Close()`? And so does `ClusterList()`?
Initially I thought that it would only be needed for those calls that do read a json from the body, etc, but it does not seem to be related.

What happens if we don't close the body?

I also fail to understand the relationship with the link you posted, please explain...

obnoxxx replied:
Also is it correct to close the body only after the block returning in case of failed status check?

obnoxxx replied:
done

obnoxxx commented:
retest this please

(libvirt/kvm cpu flags error - svm)

phlogistonjohn commented:
@obnoxxx unless @raghavendra-talur  objects I think this is probably in a good state to merge. Please rebase onto master when you have a moment.

obnoxxx commented:
@phlogistonjohn agreed, thanks.

I rebased. The only other change is that I changed the `r.Body.Close()` into a `defer r.Body.Close()`. Please see also my analysis and patches in #984.
